File: server/model/backend/ftp.js
var FtpClient = require("ftp");

// connections are reused to make things faster and avoid too much problems
const connections = {};
setInterval(() => {
    for(let key in connections){
        if(connections[key].date + (1000*120) < new Date().getTime()){
            connections[key].conn.end();
            delete connections[key];
        }
    }
}, 5000);

function connect(params){
    if(connections[JSON.stringify(params)]){
        connections[JSON.stringify(params)].date = new Date().getTime();
        return Promise.resolve(connections[JSON.stringify(params)].conn);
    }else{
        let c = new FtpClient();
        c.connect({
            host: params.hostname,
            port: params.port || 21,
            user: params.username,
            password: params.password
        });
        return new Promise((done, err) => {
            c.on('ready', function(){
                clearTimeout(timeout);
                done(c);
                connections[JSON.stringify(params)] = {
                    date: new Date().getTime(),
                    conn: c
                }
            });
            c.on('error', function(error){
                err(error)
            })
            // because of: https://github.com/mscdex/node-ftp/issues/187
            let timeout = setTimeout(() => {
                err('timeout');
            }, 5000);
        });
    }
}

module.exports = {
    test: function(params){
        return connect(params)
            .then(() => Promise.resolve(params))
    },
    cat: function(path, params){
        return connect(params)
            .then((c) => {
                return new Promise((done, err) => {
                    c.get(path, function(error, stream) {
                        if (error){ err(error); }
                        else{ done(stream); }
                    });
                });
            });
    },
    ls: function(path, params){
        return connect(params)
            .then((c) => {
                return new Promise((done, err) => {
                    c.list(path, function(error, list) {
                        if(error){ err(error) }
                        else{
                            list = list
                                .map(el => {
                                    return {
                                        size: el.size,
                                        time: new Date(el.date).getTime(),
                                        name: el.name,
                                        type: function(t){
                                            if(t === '-'){
                                                return 'file';
                                            }else if(t === 'd'){
                                                return 'directory';
                                            }else if(t === 'l'){
                                                return 'link';
                                            }
                                        }(el.type),
                                        can_read: null,
                                        can_write: null,
                                        can_delete: null,
                                        can_move: null
                                    }
                                })
                                .filter(el => {
                                    return el.name === '.' || el.name === '..' ? false : true
                                });
                            done(list);
                        }
                    })
                })
            })
    },
    write: function(path, content, params){
        return connect(params)
            .then((c) => {
                return new Promise((done, err) => {
                    c.put(content, path, function(error){
                        if (error){ err(error)}
                        else{ done('ok'); }
                    });
                });
            })
    },
    rm: function(path, params){
        return connect(params)
            .then((c) => {
                return new Promise((done, err) => {
                    c.delete(path, function(error){
                        if(error){
                            c.rmdir(path, true, function(error){
                                if(error) { err(error) }
                                else{ done('ok dir'); }
                            });
                        }
                        else{ done('ok'); }
                    });
                });
            });
    },
    mv: function(from, to, params){
        return connect(params)
            .then((c) => {
                return new Promise((done, err) => {
                    c.rename(from, to, function(error){
                        if(error){ err(error) }
                        else{ done('ok') }
                    });
                });
            });
    },
    mkdir: function(path, params){
        return connect(params)
            .then((c) => {
                return new Promise((done, err) => {
                    c.mkdir(path, function(error){
                        if(error){ err(error) }
                        else{ done('ok') }
                    });
                });
            });
    },
    touch: function(path, params){
        return connect(params)
            .then((c) => {
                return new Promise((done, err) => {
                    c.put(Buffer.from(''), path, function(error){
                        if (error){ err(error)}
                        else{ done('ok'); }
                    });
                });
            });
    }
}

File: server/model/backend/ftp.go
package backend

import (
	"fmt"
	. "github.com/mickael-kerjean/filestash/server/common"
	"github.com/secsy/goftp"
	"io"
	"os"
	"regexp"
	"strconv"
	"strings"
	"time"
)

var FtpCache AppCache

func init() {
	Backend.Register("ftp", Ftp{})

	FtpCache = NewAppCache(2, 1)
	FtpCache.OnEvict(func(key string, value interface{}) {
		c := value.(*Ftp)
		c.Close()
	})
}

type Ftp struct {
	client *goftp.Client
}

func (f Ftp) Init(params map[string]string, app *App) (IBackend, error) {
	if c := FtpCache.Get(params); c != nil {
		d := c.(*Ftp)
		return d, nil
	}
	if params["hostname"] == "" {
		params["hostname"] = "localhost"
	}
	if params["port"] == "" {
		params["port"] = "21"
	}
	if params["username"] == "" {
		params["username"] = "anonymous"
	}
	if params["username"] == "anonymous" && params["password"] == "" {
		params["password"] = "anonymous"
	}

	if params["ftp_tls"] == "implicit" {
		params["ftp_tls"] = 0
	} else {
		params["ftp_tls"] = 1
	}

	conn := 5
	if params["conn"] != "" {
		if i, err := strconv.Atoi(params["conn"]); err == nil && i > 0 {
			conn = i
		}
	}

	config := goftp.Config{
		User:               params["username"],
		Password:           params["password"],
		ConnectionsPerHost: conn,
		Timeout:            10 * time.Second,
		TLSMode:			params["ftp_tls"],
	}
	client, err := goftp.DialConfig(config, fmt.Sprintf("%s:%s", params["hostname"], params["port"]))
	if err != nil {
		return nil, err
	}
	backend := Ftp{client}

	FtpCache.Set(params, &backend)
	return backend, nil
}

func (f Ftp) LoginForm() Form {
	return Form{
		Elmnts: []FormElement{
			FormElement{
				Name:        "type",
				Type:        "hidden",
				Value:       "ftp",
			},
			FormElement{
				Name:        "hostname",
				Type:        "text",
				Placeholder: "Hostname*",
			},
			FormElement{
				Name:        "username",
				Type:        "text",
				Placeholder: "Username",
			},
			FormElement{
				Name:        "password",
				Type:        "password",
				Placeholder: "Password",
			},
			FormElement{
				Name:        "advanced",
				Type:        "enable",
				Placeholder: "Advanced",
				Target:      []string{"ftp_path", "ftp_port", "ftp_conn", "ftp_tls"},
			},
			FormElement{
				Id:          "ftp_path",
				Name:        "path",
				Type:        "text",
				Placeholder: "Path",
			},
			FormElement{
				Id:          "ftp_tls",
				Name:        "TLS mode",
				Type:        "text",
				Placeholder: "implicit",
			},
			FormElement{
				Id:          "ftp_port",
				Name:        "port",
				Type:        "number",
				Placeholder: "Port",
			},
			FormElement{
				Id:          "ftp_conn",
				Name:        "conn",
				Type:        "number",
				Placeholder: "Number of connections",
			},
		},
	}
}

func (f Ftp) Home() (string, error) {
	return f.client.Getwd()
}

func (f Ftp) Ls(path string) ([]os.FileInfo, error) {	
	return f.client.ReadDir(path)
}

func (f Ftp) Cat(path string) (io.ReadCloser, error) {
	pr, pw := io.Pipe()
	go func() {
		if err := f.client.Retrieve(path, pw); err != nil {
			pr.CloseWithError(NewError("Problem", 409))
		}
		pw.Close()
	}()
	return pr, nil
}

func (f Ftp) Mkdir(path string) error {
	_, err := f.client.Mkdir(path)
	return err
}

func (f Ftp) Rm(path string) error {
	isDirectory := func(p string) bool {
		return regexp.MustCompile(`\/$`).MatchString(p)
	}
	transformError := func(e error) error {
		// For some reasons bsftp is struggling with the library
		// sometimes returning a 200 OK
		if e == nil {
			return nil
		}
		if obj, ok := e.(goftp.Error); ok {
			if obj.Code() < 300 && obj.Code() > 0 {
				return nil
			}
		}
		return e
	}
	if isDirectory(path) {
		entries, err := f.Ls(path)
		if transformError(err) != nil {
			return err
		}
		for _, entry := range entries {
			if entry.IsDir() {
				err = f.Rm(path + entry.Name() + "/")
				if transformError(err) != nil {
					return err
				}
			} else {
				err = f.Rm(path + entry.Name())
				if transformError(err) != nil {
					return err
				}
			}
		}
		err = f.client.Rmdir(path)
		return transformError(err)
	}
	err := f.client.Delete(path)
	return transformError(err)
}

func (f Ftp) Mv(from string, to string) error {
	return f.client.Rename(from, to)
}

func (f Ftp) Touch(path string) error {
	return f.client.Store(path, strings.NewReader(""))
}

func (f Ftp) Save(path string, file io.Reader) error {
	return f.client.Store(path, file)
}

func (f Ftp) Close() error {
	return f.client.Close()
}
